app/gui2/public/visualizations/ScatterplotVisualization.vue

```diff
@@ -0,0 +1,836 @@
+  const unfilteredData = Array.isArray(rawData)
+    ? rawData.map((y, index) => ({ x: index, y }))
+    : rawData.data ?? []
+  const data: Point[] = unfilteredData.filter(
+    (point) =>
+      typeof point.x === 'number' &&
+      !Number.isNaN(point.x) &&
+      typeof point.y === 'number' &&
+      !Number.isNaN(point.y),
+  )
+  if (Array.isArray(rawData)) {
+    rawData = {}
+  }
+  const axis: AxesConfiguration = rawData.axis ?? {
+    x: { label: '', scale: Scale.linear },
+    y: { label: '', scale: Scale.linear },
+  }
+  const points = rawData.points ?? { labels: 'visible' }
+  const focus: Focus | undefined = rawData.focus
+  return { axis, points, data, focus }
+})
+
+const containerNode = ref<HTMLElement>()
+const rootNode = ref<SVGGElement>()
+const pointsNode = ref<SVGGElement>()
+const xAxisNode = ref<SVGGElement>()
+const yAxisNode = ref<SVGGElement>()
+const brushNode = ref<SVGGElement>()
+const d3Points = ref<d3Types.Selection<SVGPathElement, Point, SVGGElement, unknown>>()
+const d3Labels = ref<d3Types.Selection<SVGTextElement, Point, SVGGElement, unknown>>()
+const bounds = ref<[number, number, number, number] | null>(null)
+const brushExtent = ref<Extent | null>(null)
+const limit = ref(DEFAULT_LIMIT)
+const focus = ref<Focus>()
+
+const margin = computed(() => {
+  const xLabel = data.value.axis.x.label
+  const yLabel = data.value.axis.y.label
+  if (xLabel == null && yLabel === null) {
+    return { top: 20, right: 20, bottom: 20, left: 45 }
+  } else if (yLabel == null) {
+    return { top: 10, right: 20, bottom: 35, left: 35 }
+  } else if (xLabel == null) {
+    return { top: 20, right: 10, bottom: 20, left: 55 }
+  } else {
+    return { top: 10, right: 10, bottom: 35, left: 55 }
+  }
+})
+const width = computed(
+  () =>
+    props.width ?? props.nodeSize.x ?? containerNode.value?.getBoundingClientRect().width ?? 100,
+)
+const height = computed(
+  () => props.height ?? ((containerNode.value?.getBoundingClientRect().width ?? 100) * 3) / 4,
+)
+const boxWidth = computed(() => Math.max(0, width.value - margin.value.left - margin.value.right))
+const boxHeight = computed(() => Math.max(0, height.value - margin.value.top - margin.value.bottom))
+const xTicks = computed(() => boxWidth.value / 40)
+const yTicks = computed(() => boxHeight.value / 20)
+
+function updatePreprocessor() {
+  const args = []
+  if (bounds.value != null) {
+    args.push('[' + bounds.value.join(',') + ']')
+  } else {
+    args.push('Nothing')
+  }
+  args.push(limit.value.toString())
+  emit(
+    'update:preprocessor',
+    'Standard.Visualization.Scatter_Plot',
+    'process_to_json_text',
+    ...args,
+  )
+}
+
+onMounted(() => {
+  updatePreprocessor()
+})
+
+watchEffect(() => {
+  focus.value = data.value.focus
+})
+
+let scaleAndAxis = {} as ReturnType<typeof updateAxes>
+let zoom = {} as ReturnType<typeof addPanAndZoom>
+
+onMounted(() => {
+  scaleAndAxis = updateAxes()
+  zoom = addPanAndZoom()
```

Comment:
@Frizi i think this is the most sketchy part of the visualization code

Reply:
Combining Vue and d3 properly is kind of tricky. You usually want to keep their lands separate as much as possible. Both are responsible for doing the same thing most of the time, just have strengths in other scenarios. In this case, it would mean that vue only spawns a few root DOM nodes for d3 to pick up, and it takes it from there. You mostly did that as I just described. The tricky part is keeping that separation throuhgout the reactive and non-reactive domain as well. When something in D3 needs data from vue, you should always execute the d3 code that updates in in a `watch`, and keep in mind if you are in a non-reactive scope at that time. If you need D3 to communicate some data back to vue, you assign it to refs in d3 callbacks or directly emit if applicable. I would try to not access vue refs directly inside "d3 domain" code, as reactivity doesn't exist there. Therefore you have no "proof by construction" that D3 will be informed about it changing. Instead, whatever reactive things you need you should supply through a watch parameter.

```ts
watch(
  () => [data.value, width.value, height.value],
  ([data_, width_, height_)) => {
     // Do all d3 update stuff . Here reactive accesses are not tracked, do not access any refs in this scope.
  }
```

The remaining issue is with assigning the axes and accessing them. Afaik scales can actually be computed values, as those are pure function calls. So that makes it easier, and can effectively be a "vue domain" thing.

```ts
const xScale = computed(() => {
  return axisD3Scale(data.value.axis.x).domain(domains.value.x).range([0, boxWidth.value])
})
```

The remaining tricky part is the "entry point" - when we tell d3 to hook itself into everything?
That could be a job for a watcher on dom refs and some reactive data assigns, but I using straight up computed values where you can get away with it is the way to go.

```ts
// Make d3 selection as a computed value.
// d3.select seems to be just fine with null | undefined ¯\_(ツ)_/¯
const xAxis = computed(() => d3.select(xAxisNode.value))

// do something with selected things reactively
watch([xAxis, ticks], ([xAxis_, xTicks_, transformedXScale_) => {
  scaleAndAxis.xAxis.call(d3.axisBottom(transformedXScale_).ticks(xTicks_))
}) 
```

In general, any selection/pure computation goes into computed, and all D3 calls that do actual dom mutations go into watchers that depend on those computed. If any D3 callback needs to communicate back with reactive system, it assigns data to a ref.